docs/best-practices.rst

```diff
@@ -0,0 +1,271 @@
+        self.ti.run(ignore_ti_state=True)
+        self.assertEqual(self.ti.state, State.SUCCESS)
+        #Assert something related to tasks results
+
+Self-Checks
+------------
+
+You can also implement checks in a DAG to make sure the tasks are producing the results as expected.
+As an example, if you have a task that pushed data to S3, you can implement a check in the next task. The check should 
+make sure that the partition is created in S3 and check if the data is correct or not.
+
+Similarly, if you have a task that starts a microservice in Kubernetes or Mesos, you should check if the service has started or not using :class:`airflow.sensors.http_sensor.HttpSensor`.
+
+.. code::
+
+ task = PushToS3(...)
+ check = S3KeySensor(
+    bucket_key="s3://bucket/key/foo.parquet"
+ )
+ task.set_downstream(check)
+
+
+
+Staging environment
+--------------------
+
+Always keep a staging environment to test the complete DAG run before deploying in the production.
+Make sure your DAG is parameterized to change the variables, e.g., the output path of S3 operation or the database used to read the configuration.
+Do not hard code values inside the DAG and then change them manually according to the environment.
+
+You can use Airflow Variables to parameterize the DAG.
+
+.. code::
+
+ dest = Variable(
+    "my_dag_dest",
+    "s3://default-target/path/"
+ )
+
+Deployment in Production
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Once you have completed all the mentioned checks, it is time to deploy your DAG in production.
+To do this, first, you need to make sure that the Airflow is itself production-ready. 
+Let's see what precautions you need to take.
+
+
+Backend
+--------
+
+Airflow comes with an ``SQLite`` backend by default. It allows the user to run Airflow without any external database.
+However, such a setup is meant to be for testing purposes only. Running the default setup can lead to data loss in multiple scenarios. 
+If you want to run Airflow in production, make sure you :doc:`configure the backend <howto/initialize-database>` to be an external database such as ``MySQL`` or ``Postgres``. 
+
+You can change the backend using the following config-
+
+.. code::
+
+ [core]
+ sql_alchemy_conn = my_conn_string
+
+Once you have changed the backend, airflow needs to create all the tables required for operation.
+Create an empty DB and give airflow's user the permission to ``CREATE/ALTER`` it.
+Once that is done, you can run -
+
+.. code::
+
+ airflow upgradedb
+
+``upgradedb`` keeps track of migrations already applies, so it's safe to run as often as you need.
+
+.. note::
+ 
+ Do not use ``airflow initdb`` as it can create a lot of default connection, charts, etc. which are not required in production DB.
+
+
+Multi-Node Cluster
+-------------------
+
+Airflow uses :class:`airflow.executors.sequential_executor.SequentialExecutor` by default. It works fine in most cases. However, by its nature, the user is limited to executing at most
```

Comment:
I wouldn't say in most cases. Even on a local laptop LocalExecutor works much better.

Reply:
The problem with SequentialExecutor is it _pauses the scheduler_ when it runs a task, so even for a single node production deploy SequentialExecutor is not a good choicce.